src/Core/Services/Cache/DabCacheService.cs

```diff
@@ -78,6 +78,35 @@ public DabCacheService(IFusionCache cache, ILogger<DabCacheService>? logger, IHt
         return result;
     }
 
+    /// <summary>
+    /// Attempts to fetch response from cache. If there is a cache miss, invoke executeQueryAsync Func to get a response
+    /// </summary>
+    /// <typeparam name="TResult">Response payload Type</typeparam>
+    /// <param name="executeQueryAsync">Func with a result of type TResult. Only executed after a cache miss.</param>
+    /// <param name="queryMetadata">Metadata used to create a cache key or fetch a response from the database.</param>
+    /// <param name="cacheEntryTtl">Number of seconds the cache entry should be valid before eviction.</param>
+    /// <returns>JSON Response</returns>
+    /// <exception cref="Exception">Throws when the cache-miss factory method execution fails.</exception>
+    public async ValueTask<TResult?> GetOrSetAsync<TResult>(Func<Task<TResult>> executeQueryAsync, DatabaseQueryMetadata queryMetadata, int cacheEntryTtl)
```

Comment:
I didn't try with Jsonelement. The current Cosmos response is received as a JObject and then gets converted to a JsonDocument. I made it as TResult for two reasons:

1) To decouple the type from the Cache, so that the cache can store any data type as long as it's serializable.
2) To use the same cache method for ExecuteListAsync, which returns a list of JObject/JsonDocument. However, it later turned out that Cosmos does not support the list operation. Nevertheless, to make it future-proof, I inclined towards a generic approach instead of providing a direct type.

https://github.com/Azure/data-api-builder/blob/2ae5c0d7ab2ee03906d8fa8e70af40c086229e15/src/Core/Resolvers/CosmosQueryEngine.cs#L228C1-L229C1